WIP: add PyPi auto-publish

aaronsteers opened the pull request.

aaronsteers committed:
add PyPi auto-publish

drewbanin commented:
hey @aaronsteers - this is cool! How do you trigger the `pypi-publish` job in practice? Does that just happen on merge-to-master?

aaronsteers commented:
@drewbanin - Since I'm using github workflow and not CircleCI, I don't have an integer created by the build system which I can easily concatenate into a version number. Instead, what I do is that I create a new release in GitHub whenever I want to update the PyPi library.

My automation hook is attached to the new release 'label' and I pull the version number from the label string (e.g. `v0.9.5`, `v0.9.6`, etc.).

The actual github workflow yaml file looks like this:

```yml
name: Automated Publish (Release)

on:
  release:
    types: [published]

jobs:
  pypi_publish:
    runs-on: ubuntu-latest
    steps:
      - name: Clone git repo
        uses: actions/checkout@v1
      - name: Set up Python 3.7
        uses: actions/setup-python@v1
        with:
          python-version: 3.7
      - name: Push to PyPi (https://pypi.org/project/slalom.dataops/)
        env:
          VERSION: ${{ github.ref }}
          PYPI_USER: ${{ secrets.PYPI_USER }}
          PYPI_PASS: ${{ secrets.PYPI_SECRET }}
        run: |
          echo -e "Installing twine...\n\n"
          pip install twine
          echo -e "\nCreating setup package...\n\n"
          python setup.py sdist
          echo -e "\nPublishing to version ref '$VERSION'...\n\n"
          twine upload -u $PYPI_USER -p $PYPI_PASS dist/*
```

And actually, I remembered just now that setup.py has a role in this also:

```python
...
version_filepath = "VERSION"

if "VERSION" in os.environ:
    detected_version = os.environ["VERSION"]
    if "/" in detected_version:
        detected_version = detected_version.split("/")[-1]
if not detected_version and os.path.exists(version_filepath):
    detected_version = Path(version_filepath).read_text()
    if len(detected_version.split(".")) <= 2:
        if "BUILD_NUMBER" in os.environ:
            detected_version = f"{detected_version}.{os.environ['BUILD_NUMBER']}"
if not detected_version:
    raise RuntimeError("Error. Could not detect version.")

detected_version = detected_version.lstrip("v")
print(f"Detected version: {detected_version}")
Path(version_filepath).write_text(f"v{detected_version}")

setup(
    name="...",
    packages=["..."],
    version=detected_version,
    package_data={"": [version_filepath]},
...
```

drewbanin commented:
awesome! Thanks for sharing that with me - looks pretty neat :)
